Standard Node Material and Shader Graph

raytranuk opened the pull request:
### This PR:
- Adds private classes: `pc.StandardNodeMaterial`, `pc.ShaderNodeRegistry`, `pc.ShaderGraphBuilder` and `pc.ShaderGraphNode`
- Adds a private engine example: `shader-graph-chunks-builder` demonstrating the use of `pc.ShaderNodeRegistry` and `pc.ShaderGraphBuilder` with `pc.StandardNodeMaterial`

### example code:
```
    //create a shader graph builder
    var builder = new pc.ShaderGraphBuilder(app);

    //add input parameters
    var timeParam = builder.addParam('float', 'uTime', 0.0);
    var emissiveTexParam = builder.addParam('sampler2D', 'uEmissiveMap', emissiveMap);
    var emissiveColConst = builder.addConst('vec3', 'cEmissiveColorConst', new pc.Vec3(0.2, 0.3, 0.3));

    //create and hook up graph nodes
    var scrollNode = builder.addNode('add', timeParam, builder.addNode('uv0'));
    var emissiveSampleNode = builder.addNode('texSample', emissiveTexParam, scrollNode);
    var emissiveCalcNode = builder.addNode('mul', emissiveColConst, {node:emissiveSampleNode, port:'rgb'});

    //hook up outputs
    builder.addOutput({node:emissiveSampleNode, port:'g'}, 'float', 'sgAlpha');
    builder.addOutput(emissiveCalcNode, 'vec3', 'sgEmission');

    //return built shader graph chunk
    var shaderGraphChunk = builder.getShaderGraphChunk();
```
### Additional Implementation Notes

- In order to derive `pc.StandardNodeMaterial` and its shader generator, `pc.StandardMaterial` and the `standard` shader generator were altered - however, to try to reduce backwards compatibility issues and try to reduce future maintenance workload, the changes were kept to a minimum.
- Shader graph asset handling was removed from this PR, and will be added later alongside editor support in a Shader Graph Editor MVP PR.

I confirm I have signed the [Contributor License Agreement](https://docs.google.com/a/playcanvas.com/forms/d/1Ih69zQfJG-QDLIEpHr6CsaAs6fPORNOVnMv5nuo0cjk/viewform).

mvaligursky commented on src/scene/materials/shader-graph-node.js:
```diff
@@ -0,0 +1,614 @@
+                    if (this.graphData.ioPorts[i].type === 'sampler2D' && !this.graphData.ioPorts[i].name.startsWith('OUT_') ) {
+                        if (!(this.graphData.ioPorts[i].valueTex && this.graphData.ioPorts[i].valueTex instanceof Texture)) {
+                            return false;
+                        }
```
you should do something like this:
`var port = this.graphData.ioPorts[i];
and then access it like this:
ports.type ...
`

raytranuk replied:
I did exactly that elsewhere - not sure why I missed this loop - now fixed - thanks!

mvaligursky commented on examples/graphics/shader-graph-chunks-builder.html:
```diff
@@ -0,0 +1,134 @@
+                    app.on("update", function (dt) {
+                        time += dt;
+                        shaderGraphMaterials.forEach(function (material) {
+                            material.setShaderGraphParameter('uTime', time);
```
should we have simpler name for this instead of **setShaderGraphParameter**

raytranuk replied:
Any suggestions? This is the equivalent to setParameter - but for the shader graph embedded in the (standard node) material - so the function is setting a shader graph parameter...

mvaligursky replied:
I wonder if you could maybe even override setParameter and do some quick check to see where the parameter needs to be set?

raytranuk replied:
okay - I'll try overriding setParameter

raytranuk replied:
Overriding the setParameter function works and is definitely nicer - thanks!

mvaligursky commented on examples/graphics/shader-graph-chunks-builder.html:
```diff
@@ -0,0 +1,137 @@
+
+                    //add input parameters
+                    var timeParam = builder.addParam('float', 'uTime', 0.0);
+                    var emissiveTexParam = builder.addParam('sampler2D', 'uEmissiveMap', emissiveMap);
```
it seems the value is not optional parameter for .addParam - internally in ShaderGraphNode._addIoPort a type of it is tested and used. Do we need to be able to build a shader graph without having a value set?

mvaligursky replied:
I tried removing the emissiveMap parameter and setting it later and that seems to work. If that's supported, maybe the parameter needs to be marked as optional.

raytranuk replied:
yes - it optional - but won't be unless its set later (before rendering) - I'll mark it as optional

mvaligursky replied:
so should ShaderGraphNode._addIoPort use _type_ instead of _value_ in those ifs? Without value it falls back to what comment suggests is output port.

raytranuk replied:
I will adjust the comment

raytranuk replied:
done - made value optional when graph params (node ports) are added - and updated comment - thanks!

mvaligursky commented on src/scene/materials/shader-graph-node.js:
```diff
@@ -0,0 +1,615 @@
+        return this._addIoPort(type, 'OUT_' + name, value);
+    },
+
+    addConstant: function (type, value) {
```
curious if this could take [optional] name parameter to be used to make generated code more readable? (to be used instead of 'this.graphData.ioPorts.length' when supplied)

raytranuk replied:
yes - I can add that

raytranuk replied:
I added an optional userId - from past experience building material graphs, its rare to name constants - unless you convert it to a parameter - but this could be useful for debugging

mvaligursky commented on src/graphics/shader-node-registry.js:
```diff
@@ -0,0 +1,43 @@
+        var self = this;
+        Object.keys(coreNodeList).forEach(function (key) {
+            self._nodes[key] = new ShaderGraphNode(coreNodeList[key].code);
+            self._list.push(self._nodes[key]);
```
We should filter the duplicates from the list. Perhaps with a warning in debug mode.

mvaligursky replied:
I have not checked how this being used yet, just curious why the list is needed when you have dictionary like access with _nodes. Likely useful I'm sure, just checking.

raytranuk replied:
I will add a warning - the core nodes JSON asset should be considered an engine private asset and is invalid if core nodes are defined more than once - but being JSON format some developers will probably modify it - and a warning will also help engine development to identify and debug bad merges. The private _list is something I added for use (and is used) with the material editor prototype - and is quite useful for that.

raytranuk replied:
warning added

mvaligursky commented on examples/graphics/shader-graph-chunks-builder.html:
```diff
@@ -0,0 +1,140 @@
+                app.assets.loadFromUrl("../assets/core_nodes.json", "text", function (err, asset) {
+                    var coreNodeList = asset.resource;
+                    
+                    //register loaded core nodes
```
please use single space after // just like required in the engine code

raytranuk replied:
done

mvaligursky commented on src/scene/materials/shader-graph-builder.js:
```diff
@@ -0,0 +1,128 @@
+     * @function
+     * @name pc.ShaderGraphBuilder#addNode
+     * @description creates and adds a core node to the shader graph and connects up inputs
+     * @param {string} name - name of parameter
```
should be "name of node" I think

raytranuk replied:
yes - copy & paste typo - will fix

raytranuk replied:
thanks

mvaligursky commented on src/scene/materials/shader-graph-builder.js:
```diff
@@ -0,0 +1,128 @@
+     * @name pc.ShaderGraphBuilder#addNode
+     * @description creates and adds a core node to the shader graph and connects up inputs
+     * @param {string} name - name of parameter
+     * @param {object} [arguments] - first argument is name string, then 0-N output node ports or params to connect to new core node inputs - number and port type of arguments must match.
```
"{object} [arguments]" - I think this might create typescript definition allowing a single argument only. It might need some way to indicate list of arguments, not sure what the right way to do this is.

raytranuk replied:
I saw this format used in JSDocs in another system - I will check (google it) if it right.

raytranuk replied:
https://jsdoc.app/tags-param.html#multiple-types-and-repeatable-parameters - seems it should be {...*}

raytranuk replied:
done

mvaligursky commented on src/scene/materials/shader-graph-builder.js:
```diff
@@ -0,0 +1,128 @@
+    addNode: function (name) {
+        var args = arguments;
+        var coreNode = this._nodes[name];
+        var sgIndex = this._graph.addSubGraph(coreNode);
```
I think the public API should be more robust and check / report issues. In this case coreNode can be null, which would cause exception later. addSubGraph should handle the case of duplicate and print error / fail and similar, to help users diagnose issues.

raytranuk replied:
A check would be nice - when/if the API is public a badly typed in name could result in exceptions - what could be better is a enum of core nodes (which isn't possible yet because the complete list of core nodes hasn't yet been locked down) - for now - when used privately with the editor - its less of a problem?

raytranuk replied:
Validation has been added

mvaligursky commented on src/scene/materials/shader-graph-builder.js:
```diff
@@ -0,0 +1,128 @@
+                }
+            }
+        } else {
+            console.warn("arguments do not match core node");
```
Please make the message more informative. Node name, required and provided number of arguments.

raytranuk replied:
I will add a bit more - NB this won't be a problem for MVP - the editor's core node schemas are generated by the core node asset so they always match

raytranuk replied:
Validation - with improved error messages - has been added

mvaligursky commented on src/scene/materials/shader-graph-builder.js:
```diff
@@ -0,0 +1,128 @@
+                if (typeof(arg) === 'number') {
+                    // ret port of a node
+                    var argNodeIndex = arg;
+                    this._graph.connect(argNodeIndex, 'OUT_ret', sgIndex, sgInputs[argIndex].name);
```
I'd like this to be more robust as it's public API - this should check .connect was successful, and print errors with details if not.

raytranuk replied:
ShaderGraphNode won't be made public - additional checks could be made in the builder when/if it gets made public (after MVP).

raytranuk replied:
added validation code to the builder class

mvaligursky commented on src/scene/materials/shader-graph-builder.js:
```diff
@@ -0,0 +1,128 @@
+     */
+    addOutput: function (arg, type, name) {
+        // assumes this is only called once per graph output TODO: verify this
+        var ioPort = this._graph.addOutput(type, name);
```
yes lets verify this ;)

raytranuk replied:
Not necessary while builder is privately used by the editor - where all outputs are specified in a output node - after MVP I'd like to support custom outputs and will have to do checks for clashes at that point

mvaligursky replied:
We review this code now, and the plan is to make it public at some point, and so validation is needed. We won't be reviewing it again when we simply turn interface public, so I believe the validation needs to be done at this stage and not later.

raytranuk replied:
Fair point - I will add some validation.

raytranuk replied:
I will also say that we should be more thorough about making previously private APIs public than just removing the `@private` JSDoc keyword

raytranuk replied:
added validation code to the builder class

mvaligursky commented on src/scene/materials/shader-graph-builder.js:
```diff
@@ -0,0 +1,128 @@
+        if (typeof(arg) === 'number') {
+            // ret port of a node
+            var argNodeIndex = arg;
+            this._graph.connect(argNodeIndex, 'OUT_ret', -1, ioPort.name);
```
again - please add validation .connect was successful, and useful error message if not

raytranuk replied:
As mentioned above, used privately with the editor, it is the editor that does validation for connections - preventing invalid connections from being made - after MVP we can come up with a system to do it with code (maybe even at code build time?).

raytranuk replied:
added validation code to the builder class

mvaligursky commented on src/scene/materials/shader-graph-builder.js:
```diff
@@ -0,0 +1,128 @@
+     * @returns {any} returns the created input port
+     */
+    addParam: function (type, name, value) {
+        var ioPort = this._graph.addInput(type, name, value);
```
handle and print error when .addInput fails (duplicates, incorrect type ...)

raytranuk replied:
definitely after MVP

raytranuk replied:
added validation code to the builder class

mvaligursky commented on src/scene/materials/shader-graph-builder.js:
```diff
@@ -0,0 +1,128 @@
+     * @param {string} name - name of output
+     * @example
+     * // valid outputs are:
+     * pc.shadergraph.output(arg, 'vec3', 'vertOff');
```
I think valid outputs should be stored in a global array as data.
Something like
`var validOutputs = [ { name: "dAlpha", type: "float"}, ... ];`
This would give us:
- easier validation, and remove the need to pass in type
- would allow us to implement function to do this instead code like this being used in few places in other files:
`        if (rootShaderGraph.getIoPortByName('OUT_dAlpha') || rootShaderGraph.getIoPortByName('OUT_dNormalMap') || rootShaderGraph.getIoPortByName('OUT_dGlossiness') || rootShaderGraph.getIoPortByName('OUT_dSpecularity') || rootShaderGraph.getIoPortByName('OUT_dAlbedo') || rootShaderGraph.getIoPortByName('OUT_fragOut') || rootShaderGraph.getIoPortByName('OUT_dEmission')) {
`

raytranuk replied:
yes - when the API is made public, a public enum should be added - probably the same with the core nodes - as mentioned before, used privately with the editor there will be no mismatch - until custom outputs are added to the editor (after MVP)

raytranuk replied:
After more thought, the shader graph builder and the shader graph shouldn't the outputs - this should be done at a higher level by the system that binds the outputs - which for the MVP will the standard node material - in the future, the shader graph will be used by other systems like the particle materials and post processes. Therefore, I altered the examples given in the JSDoc

mvaligursky commented on src/scene/materials/standard-node-material.js:
```diff
@@ -0,0 +1,100 @@
+        StandardMaterial.prototype._cloneInternal.call(mat, this);
+    }
+
+    if (chunk) {
+        this._shaderGraphChunk = chunk;
+    }
```
remove the if line, so _shaderGraphChunk is always assigned to (either valid chunk, or undefined)

raytranuk replied:
true - will remove

mvaligursky commented on src/scene/materials/standard-node-material.js:
```diff
@@ -0,0 +1,100 @@
+
+Object.assign(StandardNodeMaterial.prototype, {
+
+    clone: function () {
```
this is public function (overrides Material. ) and should have documentation even if the same, otherwise it won't show up in manual

raytranuk replied:
ah - I shall put the JSDoc block back in

raytranuk replied:
these blocks are back in - but I had to mark them as `@private` because the class is private

mvaligursky commented on src/scene/materials/standard-node-material.js:
```diff
@@ -0,0 +1,100 @@
+        return clone;
+    },
+
+    setParameter: function (name, data) {
```
this is public function (overrides Material. ) and should have documentation even if the same, otherwise it won't show up in manual

raytranuk replied:
I will restore the JSDoc block

raytranuk replied:
restored - thanks!

mvaligursky commented on src/scene/materials/standard-node-material.js:
```diff
@@ -0,0 +1,100 @@
+        if (this._shaderGraphChunk) {
+            var rootShaderGraph = this._shaderGraphChunk;
+
+            var portName = 'IN_' + name + '_' + rootShaderGraph.id;
```
I wonder if there's a way to do this without allocations

raytranuk replied:
this is now lazily cached to avoid allocation here

mvaligursky commented on src/scene/materials/standard-node-material.js:
```diff
@@ -0,0 +1,100 @@
+        if (this._shaderGraphChunk) {
+            var rootShaderGraph = this._shaderGraphChunk;
+
+            rootShaderGraph.updateShaderGraphUniforms(this);
```
instead of updateShaderGraphUniforms, maybe call it updateUniforms

raytranuk replied:
done - thanks!

mvaligursky commented on examples/graphics/shader-graph-chunks-builder.html:
```diff
@@ -0,0 +1,140 @@
+                    var shaderGraphChunk = builder.getShaderGraphChunk();
+
+                    //replace model's materials with standard node materials
+                    var originalMaterials=[];
+                    var shaderGraphMaterials=[];
```
spaces around =

raytranuk replied:
done - thanks!

mvaligursky commented on src/graphics/program-lib/programs/standard-node.js:
```diff
@@ -0,0 +1,93 @@
+
+    if (options._shaderGraphChunk) {
+        graphCode[3] = '';
+        if (rootShaderGraph.getIoPortByName('OUT_dAlpha') || rootShaderGraph.getIoPortByName('OUT_dNormalMap') || rootShaderGraph.getIoPortByName('OUT_dGlossiness') || rootShaderGraph.getIoPortByName('OUT_dSpecularity') || rootShaderGraph.getIoPortByName('OUT_dAlbedo') || rootShaderGraph.getIoPortByName('OUT_fragOut') || rootShaderGraph.getIoPortByName('OUT_dEmission')) {
```
this function should be made more generic based on possible shader node outputs being specified in data, mentioned elsewhere

raytranuk replied:
I'll think about how to do this - however this code defines the chunks that gets inserted into the Standard Material's via its shader generator - so I wanted to keep it as simple as possible - with some of the same searchable terms in the standard material.

raytranuk replied:
I have improved this - I haven't turned it into a generic data driven system because it is still tied to the non-data driven standard material.

mvaligursky commented on src/graphics/program-lib/programs/standard-node.js:
```diff
@@ -0,0 +1,93 @@
+
+    var graphCode = [];
+
+    if (options._shaderGraphChunk) {
```
any reason this condition is repeated few times below, instead of all code being inside a single if statement?

raytranuk replied:
Its just a hangover from actually being embedded in the standard material shader generator - where it was useful for ironing out edge cases in the logic - I will remove the duplication - as put comments in to help maintenance when StandardMaterial is updated

raytranuk replied:
I have improved this - I haven't turned it into a generic data driven system because it is still tied to the non-data driven standard material.

mvaligursky commented on src/graphics/program-lib/programs/standard-node.js:
```diff
@@ -0,0 +1,93 @@
+
+    if (options._shaderGraphChunk) {
+        graphCode[0] = '';
+        if (rootShaderGraph.getIoPortByName('OUT_vertOff')) {
```
also this line is called again bellow, perhaps cache result, or combine those blocks?

raytranuk replied:
I have improved this - I haven't turned it into a generic data driven system because it is still tied to the non-data driven standard material.

mvaligursky commented on src/graphics/program-lib/programs/standard.js:
```diff
@@ -465,6 +467,8 @@ var standard = {
         var needsNormal = lighting || reflections || options.ambientSH || options.prefilteredCubemap || options.heightMap || options.enableGGXSpecular;
         var shadowPass = options.pass >= SHADER_SHADOW && options.pass <= 17;
 
+        needsNormal = needsNormal || graphCode;
```
does it always need normal?

raytranuk replied:
technically it doesn't - so I will add a check if the shader graph needs normal

raytranuk replied:
updated

mvaligursky commented on src/graphics/program-lib/programs/standard-node.js:
```diff
@@ -0,0 +1,93 @@
+        rootCallGLSL = rootShaderGraph.generateRootCallGlsl();
+    }
+
+    var graphCode = [];
```
instead of array which we have fixed 4 slots in that we access as array[1] and similar, make it an object with more descriptive names perhaps, making it easier on receiver side to understand the code. For example:
```
var graphCode = {};
graphCode.definitions = ...;
graphCode.functions = ...;
```

raytranuk replied:
makes sense - I will improve this

raytranuk replied:
I have improved this - I haven't turned it into a generic data driven system because it is still tied to the non-data driven standard material - and IMO it would become harder to maintain/debug.

mvaligursky commented on src/graphics/program-lib/programs/standard.js:
```diff
@@ -1417,7 +1434,36 @@ var standard = {
+
+        if (graphCode) {
+            if (!options.normalMap) {
+                code += "   dNormalMap = vec3(0);\n";
```
maybe should be 0.0 instead of 0, might generate error that int cannot be converted to float. Not sure.

raytranuk replied:
no error - GLSL allows this as constructors are used for type conversion

mvaligursky commented on src/scene/materials/shader-graph-node.js:
```diff
@@ -0,0 +1,615 @@
+import { Vec2 } from '../../math/vec2.js';
+import { Vec3 } from '../../math/vec3.js';
+import { Vec4 } from '../../math/vec4.js';
+
```
overall, I would like to see more comments in this file.

raytranuk replied:
I can do a pass and add comments to anything that seems non-obvious

raytranuk replied:
more comments added

mvaligursky commented on src/scene/materials/shader-graph-node.js:
```diff
@@ -0,0 +1,615 @@
+        for (var n = 0; n < this.graphData.ioPorts.length; n++) {
+            var ioPort = this.graphData.ioPorts[n];
+
+            if (ioPort.name.startsWith('IN_') || (ioPort.name.startsWith('CONST_') && ioPort.type === 'sampler2D')) {
```
ideally perhaps we have a list that allows us to avoid these string operations / testing here. We should have a list of uniforms to just loop over, and set parameter on material.

raytranuk replied:
yes - I will add _lists

raytranuk replied:
I've added optimizations remove/cache common string operations - however the main purpose of the class is to generate a glsl strings - so it doesn't make sense to go too far IMO

mvaligursky commented on src/scene/materials/shader-graph-node.js:
```diff
@@ -0,0 +1,615 @@
+
+                switch (ioPort.type) {
+                    case 'sampler2D':
+                        mat.setParameter(ioPort.name + chunkId, ioPort.valueTex);
```
this should be evaluated once and stored as well: ioPort.name + chunkId

raytranuk replied:
hmm - it could be but may not be much use - name is persistent, but chunkId is transient

raytranuk replied:
optimized

mvaligursky commented on src/scene/materials/shader-graph-node.js:
```diff
@@ -0,0 +1,615 @@
+        }
+    },
+
+    hasValidGraphData: function (graphRootTime) {
```
this does not have comment about the purpose and it's currently not called, so I'm not entirely sure what its purpose is - will this called when asset with graph is loaded?

raytranuk replied:
yes - this was used by the asset handler - which will dynamically swap in asset resources - probably needed for next stage of MVP

mvaligursky commented on src/scene/materials/shader-graph-node.js:
```diff
@@ -0,0 +1,615 @@
+                }
+                for (i = 0; i < this.graphData.subGraphs.length; i++) {
+                    if (this.graphData.subGraphs[i] && this.graphData.subGraphs[i] instanceof ShaderGraphNode && this.graphData.subGraphs[i].name) {
+                        if (this.graphData.subGraphs[i]._tmpRootTime === graphRootTime) {
```
it seem that the time is used as unique Id here? If yes, I think we should use some variable we keep incrementing instead, as the time has low precision (2ms on some browsers) and so will not give us unique id at all times.

raytranuk replied:
okay - I'll switch it to a counter

raytranuk replied:
done - thanks!

mvaligursky commented on src/scene/materials/shader-graph-node.js:
```diff
@@ -0,0 +1,615 @@
+            if (inOrOutAndTypeAndName[0] === "in") {
+                this.addInput(inOrOutAndTypeAndName[1], inOrOutAndTypeAndName[2]);
+            } else {
+                // unsupported parameter
```
perhaps log an error

raytranuk replied:
added a log - also revealed a small logic error - thanks!

mvaligursky commented on src/scene/materials/shader-graph-node.js:
```diff
@@ -0,0 +1,615 @@
+    },
+
+    connect: function (srcIndex, srcVarName, dstIndex, dstVarName) {
+        var connection = { srcIndex: srcIndex, srcVarName: srcVarName, dstIndex: dstIndex, dstVarName: dstVarName };
```
perhaps we could make an actual class for the connection object

raytranuk replied:
class created - does look better - thanks!

mvaligursky commented on src/scene/materials/shader-graph-node.js:
```diff
@@ -0,0 +1,615 @@
+        var ret;
+
+        if (ioPort.type === 'float') {
+            ret = 'float(' + ioPort.valueX + ')';
```
do we need to make sure values for float type contain '.' - otherwise shader compiler will see them as int and often generate incompatible type error.

mvaligursky commented on src/scene/materials/shader-graph-node.js:
```diff
@@ -0,0 +1,615 @@
+
+            depGraphList.push(funcString);
+        }
+        var tLen = depGraphList.length; // need this because pop() reduces array length!
+        for (i = 0; i < tLen; i++) {
+            generatedGlsl += depGraphList.pop();
+        }
```
while (depGraphList.length) {
    generatedGlsl += depGraphList.pop();
}

raytranuk replied:
nice - I'll swap in this code - thanks!

raytranuk replied:
done - thanks!

mvaligursky commented on src/scene/materials/shader-graph-node.js:
```diff
@@ -0,0 +1,615 @@
+        for (var i = 0; i < this.graphData.ioPorts.length; i++) {
+            var matId = '_' + this.id;
+            var ioPort = this.graphData.ioPorts[i];
+            if (ioPort.name.startsWith('IN_')) {
```
Personally I'd be tempted to make Port a class, and when its constructed with name, I would store booleans on it:
```
this.isInput = this.name.startsWith('IN_');
this.isOutput = this.name.startsWith('OUT_');
...
```

raytranuk replied:
I am considering refactoring to split out ports into categories - or maintain a transient private structure to accelerate access and/or group the split categories for convenient iteration.

mvaligursky replied:
yep either of these would work well

raytranuk replied:
I have refactored and removed prefixes entirely and replaced with PORT_TYPE_ enum - overall it is improved I think - thanks!

mvaligursky commented on src/scene/materials/shader-graph-node.js:
```diff
@@ -0,0 +1,615 @@
+                        if (dstConnectedMap[i] != undefined) {
+                            for (var n = 0; n < dstConnectedMap[i].length; n++) {
+                                var connectedSrcIndex = dstConnectedMap[i][n];
+                                if (subGraphOnListFlags[connectedSrcIndex] != true) {
```
!== instead of !=

raytranuk replied:
done - thanks!

mvaligursky commented:
I did a code review and added some inline questions / suggestion. Here are more generic points:

- can user create a new node (subgraph) that acts like a core node in a way that it can be used by other graphs? Does getShaderGraphChunk allow this perhaps?
- documentation description format (example: "@description _get shader graph chunk_") - Sentences should start with capital letter and end with full stop.
- call define SG_PS something more descriptive, what does SG stand for?
- maybe ShaderGraphBuilder.AddNode could take optional name parameter, allowing better name generation for temporary variables, making final code more readable?
- I wonder what the plan is to support WebGPU when the time comes. Are you thinking about translating glsl->webgpu shader language, or adding direct support for it inside the node system?
- how to add a constant to the graph (not a uniform, which has higher cost - cpu needs to set it every frame, takes uniform slot, storage on each material ..) - I don't see API for this, unless I missed it.

raytranuk commented:
> I did a code review and added some inline questions / suggestion. Here are more generic points:
> 
> 1. can user create a new node (subgraph) that acts like a core node in a way that it can be used by other graphs? Does getShaderGraphChunk allow this perhaps?
**Although the underlying system does support user created subgraphs, the MVP plan is not to expose them yet - to reduce editor-side work scope for MVP.**
> 2. documentation description format (example: "@description _get shader graph chunk_") - Sentences should start with capital letter and end with full stop.
**Public documentation uniformity is of course very important.** 
> 3. call define SG_PS something more descriptive, what does SG stand for?
**SG stands for Shader Graph - I can change it.**
> 4. maybe ShaderGraphBuilder.AddNode could take optional name parameter, allowing better name generation for temporary variables, making final code more readable?
**Shader code readability, unfortunately, is not something that can be maintained with most implementations of shader graphs - but after MVP I can something like this - graph readability is something that that is more possible - and there are editor features that will need to be developed after MVP**
> 5. I wonder what the plan is to support WebGPU when the time comes. Are you thinking about translating glsl->webgpu shader language, or adding direct support for it inside the node system?
**The option I was leaning toward was a WebGPU version of the Core Nodes and some minor changes to the Shader Graph systems.**
> 6. how to add a constant to the graph (not a uniform, which has higher cost - cpu needs to set it every frame, takes uniform slot, storage on each material ..) - I don't see API for this, unless I missed it.
**This was needed for the engine example so not exposed in the builder - but is supported at a lower level so will be exposed for use in the MVP editor PR.**

Thanks again for the PR review - I will do all the things I mention in my replies as promptly as I can.

raytranuk commented:
> >    **Public documentation uniformity is of course very important.**
I did a pass - should be almost ready - we can/should still do a pass when/if the docs are made public 
> >    **SG stands for Shader Graph - I can change it.**
changed to SHADERGRAPH_PS - I kept PS because it is commonly used in engine materials and shader chunks 
> >    **The option I was leaning toward was a WebGPU version of the Core Nodes and some minor changes to the Shader Graph systems.**
I've removed GLSL references in the code - at some point any remaining GLSL format strings ideally should be moved to asset files.

raytranuk commented:
> 6. how to add a constant to the graph (not a uniform, which has higher cost - cpu needs to set it every frame, takes uniform slot, storage on each material ..) - I don't see API for this, unless I missed it.

Constants are now exposed in the builder - enjoy

mvaligursky commented on src/scene/materials/shader-graph-builder.js:
```diff
@@ -0,0 +1,221 @@
+     * var node2 = builder.addNode('splitVec4', node1 );
+     * var node3 = builder.addNode('mul3', param1, node42 );
+     */
+    addNode: function (name) {
```
I wonder if this should be changed to take an array of ports to connect to inputs, instead of just list of arguments.

currently we call it
`var emissiveCalcNode = builder.addNode('mul3', emissiveColConst, {node:emissiveSampleNode, port:'rgb'});`

but we could change it to 
```
addNode: function (name, inputs)

// and call it as 
var emissiveCalcNode = builder.addNode('mul3', 
    [ emissiveColConst, {node:emissiveSampleNode, port:'rgb'} ]
 );
```

this would make it easier to read and add additional parameters at a later stage, for example:

```
addNode: function (name, inputs, precision)

// and call it as 
var emissiveCalcNode = builder.addNode('mul3', [ emissiveColConst, {node:emissiveSampleNode, port:'rgb'} ] );

var emissiveCalcNode = builder.addNode('mul3', 
    [ emissiveColConst, {node:emissiveSampleNode, port:'rgb'} ],
    pc.PRECISION_HIGH,
    ...
 );

```

It would also make it easier to do meaningful type checking using typescript.

raytranuk replied:
Thanks for this suggestion - I have done this as part of the Core Node Generators WIP commit.

raytranuk commented:
Dropping this PR back to draft to finish off Core Node Generators.

mvaligursky commented on extras/shader-nodes/texSample.js:
```diff
@@ -0,0 +1,24 @@
+texSample.meta.ports[4] = { label: 'R', id: 4 };
+texSample.meta.ports[5] = { label: 'G', id: 5 };
+texSample.meta.ports[6] = { label: 'B', id: 6 };
+texSample.meta.ports[7] = { label: 'A', id: 7 };
```
write these in mode condensed form I think. Example:

```
var texSample = {
code = "something",
meta = {
    label: "blah",
    ports = [
         { label: 'TEXTURE', id: 0 },
         { label: 'UV', id: 1 
    ]
    }
}
```

mvaligursky replied:
even one more idea would be to separate meta data that are needed for shader generation from metadata only used in editor .. so that at a later stage we can exclude them from the file. 
```
var texSample = {
code = "something",
meta = {
    ports = [.  // maybe not needed if only ids?
         { id: 0 },
         { id: 1 
    ]
    },
editor = {
    label: "blah",
    ports = [
         { label: 'TEXTURE'},
         { label: 'UV'}
    ]
}
```

raytranuk replied:
Thanks for this - I will tidy this up soon - before implementing any more core shaders nodes

raytranuk replied:
Tidied up - split up editor data from other data (explicit port ids) - should be possible to split editor data into `playcanvas-shader-nodes-editor.js` later?

mvaligursky replied:
I think add generator function directly inside, vs having to have separate cross.gen block.

I thought it could be easier to build two js files from these, and in one simply strip the editor part .. similarly to how we strip spacing and comments from shaders and code.

Either way, should be doable.

mvaligursky commented on extras/shader-nodes/add.js:
```diff
@@ -0,0 +1,17 @@
+import { flexOp } from "./flexOp";
```
for testing it's ok, but don't add them to /extras directly, create a new folder for these and add build step similar to extras to build them into separate js file.

mvaligursky replied:
see the end of rollup.config.js

raytranuk replied:
Done (adding to it extras was just for testing)

raytranuk commented:
NB Was working on dynamic switches - which also benefit from the Core Node Generator flexibility of having N arguments - when I realized I had all the pieces to implement static switches - so I did - and they are functional. Will test more - but looks like Render Pass, Platform/API/Extension, Quality Setting and user defined static switches will work.

raytranuk commented:
After implementing the flexible operator and flexible switch in core node generation - it seemed that adding flexibility into all connections would also be possible - so I have added:
- per connection swizzle - just specify the swizzle when passing parameters to addNodes()
- automatic channel truncation/extension/broadcast - this works with or without swizzle

mvaligursky commented on src/scene/materials/standard-node-material.js:
```diff
@@ -0,0 +1,189 @@
+        if (this.shaderGraphChunk) {
+            if (restoreGraphDefaults === true) {
+                // restore
+                this.shaderGraphChunk.setParameters(device, Object.keys(this._graphParamOverrides), null);
```
Object.keys allocates array, which is not ideal. We should either store overrides in an array directly, or just pass the object as is and iterate over it.

raytranuk replied:
Good point - maintaining an array alongside the object makes the most sense.

raytranuk replied:
refactored

mvaligursky commented on src/scene/materials/standard-node-material.js:
```diff
@@ -0,0 +1,189 @@
+        return this._shaderGraphChunk;
+    },
+    set: function (value) {
+        if (!value || this._shaderGraphChunk === value)
```
just curious about the !value here. Should the code that removes current chunk execute? When I set something to null, I'd expect current value to be removed.

raytranuk replied:
yeah not sure why its here - I shall remove it

raytranuk replied:
removed

mvaligursky commented on src/scene/materials/standard-node-material.js:
```diff
@@ -0,0 +1,189 @@
+        // remove this material from current chunk's material list
+        if (this._shaderGraphChunk) {
+            var self = this;
+            this._shaderGraphChunk._materials = this._shaderGraphChunk._materials.filter(function (mat) {
```
ideally we do this in place without extra array allocation

raytranuk replied:
I will change it to splice()

raytranuk replied:
refactored - thanks!

mvaligursky commented on src/scene/materials/shader-graph-builder.js:
```diff
@@ -0,0 +1,350 @@
+     * @function
+     * @name pc.ShaderGraphBuilder#getShaderGraphChunk
+     * @description Get shader graph chunk.
+     * @returns {any} Shader graph chunk.
```
should the output type be a real type rather then any?

raytranuk replied:
It should be `ShaderGraphNode` - I'll update it

raytranuk replied:
updated

mvaligursky commented on src/scene/materials/standard-node-material.js:
```diff
@@ -0,0 +1,189 @@
+ * @classdesc StandardNodeMaterial is sub class of the StandardMaterial, and adds shader graph
+ * interop functionality.
+ * @param {pc.StandardMaterial} mat - Optional material which is cloned.
+ * @param {any} chunk - Optional shader graph node chunk to be used.
```
does it take "any" parameter?

raytranuk replied:
It should be ShaderGraphNode - I'll update it

raytranuk replied:
updated

mvaligursky commented on src/scene/materials/shader-graph-node.js:
```diff
@@ -0,0 +1,860 @@
+    },
+
+    getIoPortByName: function (name) {
+        return (this._portMap[name] === undefined) ? undefined : this.graphData.ioPorts[this._portMap[name]];
```
this._portMap[name] should not execute twice

raytranuk replied:
I will add a var

raytranuk replied:
improved

mvaligursky commented on src/scene/materials/shader-graph-node.js:
```diff
@@ -0,0 +1,860 @@
+    },
+
+    getIoPortUniformName: function (name) {
+        return (this._portMap[name] === undefined) ? undefined : this._getCachedPort(this._portMap[name]).nameId;
```
same here, this._portMap[name] should not execute twice

raytranuk replied:
I will add a var

raytranuk replied:
improved

mvaligursky commented on src/scene/materials/shader-graph-node.js:
```diff
@@ -0,0 +1,860 @@
+
+    _getCachedPort: function (i) {
+        var ioPort = this.graphData.ioPorts[i];
+        if (!this._portCache[ioPort.name]) this._portCache[ioPort.name] = { isUniform: (ioPort.ptype === PORT_TYPE_IN || (ioPort.ptype === PORT_TYPE_CONST && ioPort.type === 'sampler2D')), nameId: ioPort.name + '_' + this.id, port: ioPort, index: i };
```
make it a multiline :)

raytranuk replied:
okay

raytranuk replied:
changed

mvaligursky commented on examples/graphics/shader-graph-chunks-overrides.html:
```diff
@@ -0,0 +1,215 @@
+
+                        // add input parameters
+                        var timeParam = builder.addParam('float', 'uTime', 0.0);
+                        var texParamA = builder.addParam('sampler2D', 'uTexMapA', logoMap);
```
the value is described as optional (and should be optional), but crashes when not specified.

mvaligursky commented on examples/graphics/shader-graph-chunks-builder.html:
```diff
@@ -0,0 +1,176 @@
+
+                    // add input parameters
+                    var timeParam = builder.addParam('float', 'uTime', 0.0);
+                    var texParamA = builder.addParam('sampler2D', 'uTexMapA', cloudsMap);
```
the value (cloudsMap) is described as optional (and should be optional), but crashes when not specified.

raytranuk replied:
hmm - probably best to keep it optional, and have the builder put in a placeholder if not supplied - I'll see if I can add a placeholder (this is what materials do (in the material handler) I think)

raytranuk replied:
refactored - Texture parameters are now treated separately - with additional validation and placeholder.

mvaligursky commented on examples/graphics/shader-graph-chunks-builder.html:
```diff
@@ -0,0 +1,176 @@
+                                material.overrideGraphSwitch(switchName, switchValue);
+                                material.update();
+                            }
+                            material.overrideGraphParameter('uTime', time);
```
I'd prefer these would be set as any other material properties using setParameter to make it simpler for users. A programmer might not know if it is parameter or override, and that can be changed and code would need to be updated.

raytranuk replied:
If using the same function name it wouldn't be possible to validate and give warnings if the parameter exists on the graph or not - which could be useful to developers - it also could lead to bugs that occur if the intention wasn't to override a parameter or switch, leading to odd behavior or performance - all of this can also be said of the mesh instance overrides which do use the same function name - so at least it will be consistently tricky to debug :)

mvaligursky replied:
well a user of a shader does not need to understand internals of the material / graph implementation .. in the future they get a shader graph material from a store, or somebody else on the team builds it, and they just need to set parameters.

raytranuk replied:
> small comment, but I'll bring it up just for consideration. These are a few lines of generated shader:
> 
> ```
> vec2 temp_vec2_0;
> vec2 temp_vec2_1;
> temp_vec2_0 = uv0(  );
> temp_vec2_1 = add( float(uTime), vec2(temp_vec2_0) );
> ```
> 
> could this be generated instead? For readability:
> 
> ```
> vec2 temp_vec2_0 = uv0(  );
> vec2 temp_vec2_1 = add( float(uTime), vec2(temp_vec2_0) );
> ```

It could - I would have to add a flag of some sort to indicate the output is a return value - but certainly doable - and make the shader a bit shorter

raytranuk replied:
I've made general parameters use setParameters - but I am treating switches differently - because they are different.

mvaligursky commented on shader-nodes/shader-nodes/cross.js:
```diff
@@ -0,0 +1,37 @@
+cross.gen = function ( argTypes, options ) {
+    var code = this.code;
+
+    if (options && options.precision) {
+        // precision - alter name to create variant
+        code = code.replace('vec3 cross', 'vec3 cross_' + options.precision);
+        // precision - tmp: useful comment
+        code = code.replace( '{\n', '{\n// precision ' + options.precision + ' float;\n');
+    }
```
I think you should try something more elegant than starting with a code string and then applying some hard to read string.replace calls.

Maybe even template strings directly (see https://hacks.mozilla.org/2015/05/es6-in-depth-template-strings-2/) - you'd need to merge master to your branch to get ES6 support. Or something along similar format to allow easy replacements are runtime, and good readability.

(adding as a screenshot to get better highlighting)
<img width="370" alt="Screen Shot 2020-12-31 at 2 29 53 PM" src="https://user-images.githubusercontent.com/59932779/103414165-aca95480-4b74-11eb-8d61-1b4f44e069e2.png">

raytranuk replied:
ooh nice - I'll merge master and use this - we can also improve a lot of the shader chunk stuff in the generator

raytranuk replied:
I'm planning to look into this - but is it necessary for this PR? This PR isn't supposed to contain the final list of core nodes or core node generators - just the underlying systems to allow for them to be developed (alongside the MVP graph editor). In fact it would be great (and probably necessary) if core nodes could be developed by several developers and eventually by external contributors

mvaligursky replied:
My opinion is that any code that makes it into code base needs to be a good quality, as we wont be reviewing it again. I agree all core nodes can be developed by multiple programmers at a later stage, but here we're looking at a system that allows us to do it effectively. But I understand that it's not easy to develop everything in a larger PR to best quality, and so I'll call @willeastcott for an opinion here as well.

raytranuk replied:
IMO replace() isn't low quality code - it is what was possible without newer es6 language features - and is used in the engine in the shader generators for all material texture map UVs.

raytranuk replied:
I've replaced `replace()` (and some other string generation) with string templates - it does increase readability a bit.

mvaligursky commented on src/scene/materials/shader-graph-builder.js:
```diff
@@ -0,0 +1,350 @@
+                var core = this._addedNodes[arg.node];
+                if (!core) {
+                    console.error(err + ": invalid node index: " + arg.node);
+                    return false;
```
the function returns an array .. so for type consistency, this should return null instead of false.

raytranuk replied:
I'll change it

raytranuk replied:
done

mvaligursky commented:
small comment, but I'll bring it up just for consideration. This is few lines of generated shader:

```
vec2 temp_vec2_0;
vec2 temp_vec2_1;
temp_vec2_0 = uv0(  );
temp_vec2_1 = add( float(uTime), vec2(temp_vec2_0) );
```

could this be generated instead? For readability:

```
vec2 temp_vec2_0 = uv0(  );
vec2 temp_vec2_1 = add( float(uTime), vec2(temp_vec2_0) );
```

mvaligursky commented on src/graphics/shader-node-registry.js:
```diff
@@ -0,0 +1,88 @@
+};
+
+ShaderNodeRegistry.prototype.get = function (name, argTypes, options) {
+    if (this._nodeCache[name]) {
```
you should store the result of this instead of executing it many times, as each of these is a dictionary lookup.
`this._nodeCache[name]`

raytranuk replied:
Not sure what you mean here - the node registry is just a single instance per app - and this core node cache is only used when building graphs - which happens only on shader graph editing/building or generation.

mvaligursky replied:
What I meant what that instead of calling this many times:
`this._nodeCache[name]`
ideally you store it in a var and reuse:
`var namedNode = this._nodeCache[name];`

raytranuk replied:
ah - okay - good point - and done.

mvaligursky commented on src/graphics/shader-node-registry.js:
```diff
@@ -0,0 +1,88 @@
+    if (options) {
+        var optionKeys = Object.keys(options);
+        for (var keyIndex = 0; keyIndex < optionKeys.length; keyIndex++) {
+            variantKey += '_' + optionKeys + '_' + options[optionKeys[keyIndex]];
```
I think `+ optionKeys +` is not what you want there, as optionKeys is an array and that add all array keys there, should be `+ optionKeys[keyIndex] +` I suspect?
There is probably better way to walk this anyways, without calling `Object.keys` which creates an array, and without having to do lookup on each step `options[optionKeys[keyIndex]]`

raytranuk replied:
yes - it was a typo - refactored to fix this and be more efficient - and looks simpler

raytranuk commented:
> small comment, but I'll bring it up just for consideration. These are a few lines of generated shader:
> 
> ```
> vec2 temp_vec2_0;
> vec2 temp_vec2_1;
> temp_vec2_0 = uv0(  );
> temp_vec2_1 = add( float(uTime), vec2(temp_vec2_0) );
> ```
> 
> could this be generated instead? For readability:
> 
> ```
> vec2 temp_vec2_0 = uv0(  );
> vec2 temp_vec2_1 = add( float(uTime), vec2(temp_vec2_0) );
> ```

Should be doable - would need to add a flag to indicate the output is a return value.

mvaligursky commented on src/scene/materials/shader-graph-builder.js:
```diff
@@ -0,0 +1,393 @@
+    this._graph.name = 'graphRoot_' + id;
+
+    // for validation
+    this._addedNamedInputs = {};
+    this._addedNodes = [];
+    this._addedOutputs = {};
```
curious why these names start with 'added'?
Maybe just _namedInputs, _nodes, _outputs?

raytranuk replied:
I guess they can be renamed - they are just private data used for validation - maybe they can be tidied up by putting them into a _validationData struct?

mvaligursky replied:
good idea.

mvaligursky commented on src/scene/materials/shader-graph-builder.js:
```diff
@@ -0,0 +1,393 @@
+    this._placeholderTex = _placeholderTex;
+};
+
+ShaderGraphBuilder.prototype.constructor = ShaderGraphBuilder;
```
is this line needed or just some leftover from when there was different base class?

raytranuk replied:
I think it was indeed a leftover - removing

mvaligursky commented on src/scene/materials/shader-graph-builder.js:
```diff
@@ -0,0 +1,393 @@
+        _placeholderTex.name = 'placeholder';
+
+        // fill pixels with grey
+        var pixels = _placeholderTex.lock();
+        for (var i = 0; i < 4; i++) {
+            for (var c = 0; c < 4; c++) {
+                pixels[i * 4 + c] = [128, 128, 128, 255];
+            }
+        }
```
This does not fill the texture with those pixels .. you're assigning an array of 4 elements to each pixel, that does do what you need. The final pixels array is just full of zeros.
Maybe look .fill, as you get typedArray from .lock()
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/TypedArray/fill

raytranuk replied:
Ah yes - made a mistake in adjusting this code when copying from the asset handler - I will switch to fill - thanks!

mvaligursky commented on src/scene/materials/shader-graph-builder.js:
```diff
@@ -0,0 +1,393 @@
+        _placeholderTex.unlock();
+    }
+
+    this._placeholderTex = _placeholderTex;
```
I think don't assign _placeholderTex to this._placeholderTex and just use _placeholderTex directly when you need to access it. It's static, and so no need to assign it to instance of the class.

raytranuk replied:
ok - will do

mvaligursky commented on shader-nodes/shader-nodes/flexOp.js:
```diff
@@ -0,0 +1,65 @@
+            // convert all arguments to return type
+            code += `${retType}(arg${argIndex}`;
+
+            // expand if needed - extra components are set to zero - NB scalar values are broadcast to all components
+            var argComp = type2Comp[argTypes[argIndex]];
+            var extend = type2Comp[retType] - argComp;
+            code += (extend > 0 && argComp !== 1) ? `, ${comp2Type[expand]}(0))` : ')';
+
+            code += (argIndex === argTypes.length - 1) ? ';\n' : ` ${opCode} `;
```
it feels like this should be on a higher level when the generated node function gets called .. so that it works any node parameters (and not only flex nodes), without having to implement this for each core node / subgraph?

raytranuk replied:
Auto type conversion actually already does work on all nodes in the graph - in combination with (an optional) per connection swizzle - which makes it very flexible - so you can feed a scalar value into vector port or vice versa or a vec2 into a vec3 or a vec4 into a vec3 etc and it gets automatically extended or truncated: https://github.com/playcanvas/engine/blob/3b7e6dbf40c1f1e1a844234e5dd9579da712cf7e/src/scene/materials/shader-graph-node.js#L423 - its pretty easy to try in the engine examples - I can also add one that demos Auto type conversion? It should also become more intuitive when implemented in the graph editor with colors (like in Lens Studio) and/or numbers (link in Unity) indicating the type conversion.

The flexOp is actually distinct from general auto type conversion because it and can take a flexible number of arguments of any (scalar or vector) type and changes the type of the operation itself and based on the argument with the largest number of components - this means other arguments may need to be extended.
